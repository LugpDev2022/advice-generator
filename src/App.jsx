import React, { useEffect, useState } from "react";
import { Col, Container, Modal, Row, Spinner } from "react-bootstrap";
import desktopDivider from "./assets/images/pattern-divider-desktop.svg";
import mobileDivider from "./assets/images/pattern-divider-mobile.svg";
import iconDice from "./assets/images/icon-dice.svg";

function App() {
  const [{ loading, advice }, setState] = useState({
    loading: true,
    advice: {},
  });

  const getAdvice = () => {
    fetch("https://api.adviceslip.com/advice")
      .then((resp) => {
        return resp.json();
      })
      .then((data) => {
        setState({
          loading: false,
          advice: data,
        });
      });
  };

  useEffect(() => {
    getAdvice();
  }, []);

  const handleClick = () => {
    setState({
      loading: true,
      advice,
    });

    getAdvice();
  };

  if (loading) {
    return (
      <Spinner
        animation="border"
        variant="primary"
        role="status"
        className="position-absolute top-50 start-50 "
      >
        <span className="visually-hidden">Loading...</span>
      </Spinner>
    );
  }

  return (
    <Modal
      show
      animation={false}
      backdrop={false}
      centered
      className="my-modal"
    >
      <Modal.Body>
        <Container>
          <Row>
            <Col>
              <h1 className="custom-title text-info text-center my-3">
                ADVICE #{advice.slip.id}
              </h1>
            </Col>
          </Row>
          <Row>
            <Col>
              <p className="custom-p text-center text-primary">
                {advice.slip.advice}
              </p>
            </Col>
          </Row>

          <Row>
            <Col className="justify-content-center d-flex mt-2 mb-5 overflow-hidden">
              <picture>
                <source srcSet={mobileDivider} media="(max-width: 380px)" />
                <img src={desktopDivider} />
              </picture>
            </Col>
          </Row>

          <Row>
            <Col>
              <button
                className="custom-btn rounded-circle"
                onClick={handleClick}
                disabled={true}
              >
                <img src={iconDice} />
              </button>
            </Col>
          </Row>
        </Container>
      </Modal.Body>
    </Modal>
  );
}

export default App;
