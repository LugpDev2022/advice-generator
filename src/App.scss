@font-face {
  font-family: "Manrope";
  src: url("./assets/fonts/Manrope.ttf");
}

* {
  font-family: "Manrope";
  font-weight: 800;
}

body {
  background: #1f2632 !important;
}

$theme-colors: (
  "primary": #cee3e9,
  "secondary": #323a49,
  "info": #52ffa8,
  "light": #4e5d73,
);

@import "bootstrap/scss/bootstrap";
