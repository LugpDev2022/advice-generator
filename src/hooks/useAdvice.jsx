import { useEffect, useState } from 'react';

export const useAdvice = () => {
  const [state, setState] = useState({
    loading: true,
    canSearch: false,
    advice: {},
  });

  useEffect(() => {
    fetch('https://api.adviceslip.com/advice', { cache: 'no-store' })
      .then((resp) => {
        return resp.json();
      })
      .then((data) => {
        setState({
          loading: false,
          canSearch: false,
          advice: data,
        });

        setTimeout(() => {
          setState((prevState) => {
            return {
              ...prevState,
              canSearch: true,
            };
          });
        }, 2500);
      });
  }, []);

  const handleClick = () => {
    setState({
      ...state,
      loading: true,
    });

    fetch('https://api.adviceslip.com/advice', { cache: 'no-store' })
      .then((resp) => {
        return resp.json();
      })
      .then((data) => {
        setState({
          loading: false,
          canSearch: false,
          advice: data,
        });

        setTimeout(() => {
          setState((prevState) => {
            return {
              ...prevState,
              canSearch: true,
            };
          });
        }, 2500);
      });
  };

  return { ...state, handleClick };
};
