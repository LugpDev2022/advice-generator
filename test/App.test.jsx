import React from "react";
import { render, screen } from "@testing-library/react";
import App from "../src/App";
import { useAdvice } from "../src/hooks/useAdvice";

describe("<App />", () => {
  test("should match snapshot", () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  test("should show the spinner", () => {
    useAdvice.mockReturnValue({ loading: true });
    render(<App />);

    screen.debug();
  });
});
