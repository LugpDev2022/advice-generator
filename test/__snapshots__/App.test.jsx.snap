// Vitest Snapshot v1

exports[`<App /> > should match snapshot 1`] = `
<div>
  <div
    class="position-absolute top-50 start-50  spinner-border text-primary"
    role="status"
  >
    <span
      class="visually-hidden"
    >
      Loading...
    </span>
  </div>
</div>
`;
